server/src/main/java/org/opensearch/gateway/GatewayAllocator.java

```diff
@@ -303,6 +313,59 @@ protected AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShards.Nod
         }
     }
 
+
+    class InternalPrimaryBatchShardAllocator extends PrimaryShardBatchAllocator {
+
+        @Override
+        @SuppressWarnings("unchecked")
+        protected AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShardsBatch> fetchData(Set<ShardRouting> shardsEligibleForFetch,
+                                                                                                                                   Set<ShardRouting> inEligibleShards,
+                                                                                                                                   RoutingAllocation allocation) {
+            ShardRouting shardRouting = shardsEligibleForFetch.iterator().hasNext() ? shardsEligibleForFetch.iterator().next() : null;
+            shardRouting = shardRouting == null && inEligibleShards.iterator().hasNext() ? inEligibleShards.iterator().next() : shardRouting;
+            if (shardRouting == null) {
+                return new AsyncShardFetch.FetchResult<>(null, Collections.emptyMap());
+            }
+            String batchId = getBatchId(shardRouting, shardRouting.primary());
+            if (batchId == null) {
+                logger.debug("Shard {} has no batch id", shardRouting);
+                throw new IllegalStateException("Shard " + shardRouting + " has no batch id. Shard should batched before fetching");
+            }
+
+
+            if (batchIdToStartedShardBatch.containsKey(batchId) == false) {
+                logger.debug("Batch {} has no started shard batch", batchId);
+                throw new IllegalStateException("Batch " + batchId + " has no started shard batch");
+            }
+
+            ShardsBatch shardsBatch = batchIdToStartedShardBatch.get(batchId);
+            // remove in eligible shards which allocator is not responsible for
+            inEligibleShards.forEach(GatewayAllocator.this::safelyRemoveShardFromBatch);
+
+            if (shardsBatch.getBatchedShards().isEmpty() && shardsEligibleForFetch.isEmpty()) {
+                logger.debug("Batch {} is empty", batchId);
+                return new AsyncShardFetch.FetchResult<>(null, Collections.emptyMap());
+            }
+
+            Map<ShardId, Set<String>> shardToIgnoreNodes = new HashMap<>();
+
+            for (ShardId shardId : shardsBatch.asyncBatch.shardToCustomDataPath.keySet()) {
+                shardToIgnoreNodes.put(shardId, allocation.getIgnoreNodes(shardId));
```

Comment:
Can the `shardToIgnoreNodes` map have empty (set) values ? Can we ignore adding the entry in such cases? 

Ref: 

https://github.com/opensearch-project/OpenSearch/blob/f7f3500cc16516d4ac44e10bc333529881ad5762/server/src/main/java/org/opensearch/cluster/routing/allocation/RoutingAllocation.java#L241-L248

Reply:
Even if we ignore it will be later created by AsyncShardFetch object for completeness sake.

Reply:
Can you link the code where we are adding the entries with empty set. Is there scope to optimize here - Avoid creating the empty sets that server no purpose ?

Reply:
Can you link the code where we are adding the entries with empty set. 


Is there scope to optimize here - Avoid creating the empty sets that server no purpose ?